### ISSUE TITLE
Create new members upon registration using events.

### ISSUE THREAD

**Issue by pameck** [rabblerouser/core#112]
This is the first step into our event-sourcing adventure! 

We need to refactor the current backend flow for creating members upon registration.

When a member signs up, an event should be generated, then another service will listen to that event and create a new member record on the existing DB.

![whatsapp image 2017-01-07 at 4 59 18 pm](https://cloud.githubusercontent.com/assets/3664374/21739822/e537ad0a-d4fa-11e6-93c1-d07d45a9d53e.jpeg)

**Considerations**: 

1. Abstract the event specific code in a way that we can switch the event-related technology if necessary. 
1. AWS Kinesis is our choice (unless there is something more appropriate)
1. The component that transforms the event into a member could be a module inside the existing app, or could be any other thing.

**Comment by pameck** [comment]
WIP:
## Things to consider

1. **Sync vs Async actions:**
When a user is signing up, we need to consider that they are waiting for a response (synchronously). Events are async so we need to figure out how to synchronously respond to the user, some thoughts: 

    - Do the unique-id validation (the light-weight stuff) synchronously, and then leave the saving,   emailing, etc. (heavy-weight stuff) to kinesis and friends
    - Have the UI poll until it gets a result. (ugh!)

1. **`newUserRequested` vs `newUserCreated`**
One thing is to listen to the event sent upon user sign up, and another thing is to listed to the event once the user has been successfully created (a.k.a persisted in DB?)

1. What happens if the service processing the **`newUserRequested`** event is down? (How do we retry?)

1. What happens if the same event is read by a service more than once?

1. What happens if, for whatever reason, there are consecutive events that don't get read in the logical order?

**Comment by yearofthedan** [comment]
Just on point 1. Is the unique id concern an issue yet?  Members won't be logging in to the app, and there is a story which (we touched very quickly on Saturday re this) wants to resolve the same member who registers with different emails. So the primary key may not be driven from the front end for members.
Can we rely on the front-end validation, and tell the user it went well if the FE validation passes and the event went onto the queue?

**Comment by yearofthedan** [comment]
Here's the story: https://github.com/rabblerouser/rabblerouser-core/issues/103

**Comment by pameck** [comment]
@yearofthedan what I meant with point 1 was more generic, as in stating the fact that there might be things we need to do synchronously. If, in this case, everything can be handled by the FE that'd great, but if it doesn't we may have to consider splitting asycn and sync actions. 

I haven't looked into the flow in that much detail yet, but I'll keep in mind story #103 to see if it is easier to play it at the same time.

**Comment by camjackson** [comment]
1. I think sync vs. async stuff is going to be a common theme in trying to do everything with event sourcing. For something like this, I think you have to treat writing to the stream the same as writing to a traditional database. So once the backend writes the registration event to the stream (which would happen after any validation), it can return a success response to the frontend. All the stuff that happens after that (sending the email, copying the data to the SQL db) is up to the other backend stuff to figure out, and the frontend shouldn't be concerned with it. As far as the frontend <-> backend interaction is concerned, if the registration event is written to the stream, then the user has been created. Also for unique IDs it feels like we could just the server generate random UUIDs for each request?
2. Do we need different events for those things at the moment? (See below...)
3. (and also 4.) Yeah these are hard solve. Kinesis may have some of that stuff builtin? I.e. Event Store has a thing called "competing consumers" where multiple clients can be subscribed to a particular event, but only one of them will actually receive it. Then once that client has successfully processed that event, it sends an ACK back to the store to let it know. If the ACK never comes back (or if it gets a NACK), then it puts the event back out again to a different consumer. This should guarantee that the event is processed (successfully) exactly once. Needs investigating to see if Kinesis Streams supports competing consumers.
4. (see above)
5. For registration, do we care? We can probably solve this later for things that are order-sensitive.

Here is roughly how I would structure it:
1. Frontend POSTs the data to the signup backend (right now it's all the same backend, but it could be separate)
2. Signup backend does synchronous validation, with a 400 if it fails
3. Signup backend puts a `memberRegistered` event on the stream, which as far as it is concerned constitutes success/creation.
3. Signup backend returns a 200/201/202 to the frontend.
4. There are two side effects to the registration event:
    1. Creating a permanent record of that user (because Kinesis is the comms, not the persistence)
        - In the future this would just mean writing the event to an event store
        - Right now it means updating the SQL database with the new user
        - That data will then be read when there are future queries for the member list
    2. Sending an email:
        - I guess this is just some stateless function that reads the event and sends the email
        - Ideally use a competing consumers system to make sure this happens exactly once

The 1st side effect (persistence) will always be just a single thing that updates the permanent data store. The only thing that changes there is whether we're writing it to a sql store or an event store. For the 2nd one though, there could be any number of independent services watching for that event. E.g. maybe in the future we send an email to the new member, and we send an email to the branch admin, and we also tweet that someone just joined our organisation.

This is all just throwing ideas out there, I've never actually tried to do this stuff before. Thoughts?

**Comment by pameck** [comment]
@yearofthedan and I had some sort of story kick-off and these are the tasks we will be working on:

## Infra 
- [x] Setup a Kinesis Stream 
* Includes investigation to whether _AWS Firehose_ is a proper alternative.
* Add its creation to our infra repo 

## Sign up Flow
- [x] Have the sign-up app write to the stream
* Isolate this as much as possible (we might pull out the sign-up app in the future).
* extract the producer into its own `npm module`.
* Keep writing to the DB, once we have the consumer ready, we will plug it off. 

- [x] Remove DB write
* Once the DB consumer is ready.

- [x]  DB consumer
* Reads from the stream and creates the new member
* Writes a `memberCreated` or `memberCreationFailed` event to the stream 
* Extract the kinesis consumer into its own module (to encapsulate all the `AWS` logic)

## Confirmation Email Flow

- [x] Email consumer
* Listens to `memberCreated` and sends an email.
* We need to make sure only one confirmation email is sent upon member creation.

- [x] Remove current email service
* Once the Email consumer is ready



**We shouldn't have any `aws-sdk` related to `kinensis` in our code if we did this correctly.**

**Comment by camjackson** [comment]
Sounds like a pretty solid plan! One question: are we going to do anything with the `memberCreated` and `memberCreationFailed` events yet?

**Comment by pameck** [comment]
@camjackson The idea is for the confirmation email to be sent upon `memberCreated` instead of `memberRegistered`. 

1. The sign-up app produces `memberRegistered`
1. The db "companion thingy" consumes `memberRegistered`, saves the new member to the DB and produces `memberCreated`
1. The email service consumes `memberCreated` and sends the confirmation email.

As for `memberCreationFailed`, not sure, we haven't gone that deep into the story, it might not even be necessary.

**Comment by camjackson** [comment]
Hmmm. I'm unsure about the need for the `memberCreated` event. To me, the registration event kind of already *is* the event that says that the member has been created. I get that technically right now that's not true, because we still have to have the record be persisted to the relational database, but my understanding is that in event-sourcing-land, the fact that the registration event got put on the stream, means that the user is created - no confirmation needed.

**Comment by yearofthedan** [comment]
I may be a bit unclear, but something has to read the event and persist it
right, even if it's some sort of event store? So it's only once it's in the
store that the member is created and hence an event.

I'm just thinking of whether there could be issues if the persistence is
down or if for some reason the event can't be stored (invalid?).

I think some of this will become clearer as we build it.

**Comment by camjackson** [comment]
Yeah, there has to be some mechanism of persisting the event from the stream. In the future we should be able to pretty easily pipe the stream into e.g. an S3 bucket (this may or may not require something like a lambda function in the middle to join the two up). But yeah right now it's all a bit more unreliable, because we have to keep the SQL db in sync, which is a pain.

It's a good point about stuff being down. I'm just not sure how much the rest of the whole system should be concerned with the event persistence. I.e. should the thing that triggers emails upon registration be concerned with whether the event store writing thing is currently up? Or should it treat the registration event as the source of truth, and assume that the persistence will eventually be sorted out by someone else, through retries etc.

I feel like if we can't trust our domain events coming down the stream, then we're going to end up with double events for every single interaction -- one to say that it's happening, and one to say that it's happened.

All of that said, I'm wary of over-thinking it all too much before we've started building it. So I think @yearofthedan and @pameck should go ahead with whatever you think makes sense, and then we can answer questions like this when we know more :)

**Comment by pameck** [comment]
After going through the AWS docs for the events consumer, @yearofthedan and I found this

> Your application should have one thread per shard, each reading continuously from its stream. To read from a stream continually, call GetRecords in a loop.” - [see here](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_GetRecords.html)

We were planning to have an `npm` package that we could import in our single-threaded NodeJS Rabble Rouser core to listen to events. But the NodeJS became a limitation.  AWS's solution for its NodeJS client was to have a runtime dependency on a Java component that listens for events and calls the JS client. 

We decided not to have a runtime dependency with Java and instead use a Lambda to listen to the events and call Rabble Rouser core. The attached image reflects our plan after our findings. 

<img width="753" alt="screen shot 2017-01-19 at 10 44 08 am" src="https://cloud.githubusercontent.com/assets/3664374/22094243/1f06cd90-de60-11e6-932f-9229351203aa.png">
